bolt11: use custom mutator

brunoerg opened the pull request:
Fixes #120 

This PR adds the custom mutator from core lightning's bolt11 target to the `deserialize_invoice` target here. I did some modifications due to C++ and the bech32 functions used are from Bitcoin Core (so it's required to build bitcoin core module).

brunoerg committed:
bolt11: use custom mutator

brunoerg committed:
ci: add -DCUSTOM_MUTATOR_BOLT11

brunoerg commented:
cc @morehouse @erickcestari

erickcestari commented:
I am having three issues, may some of these are not related to this pr:

1. When trying to build it, I got some errors.
```
make
clang++ -DLDK -DLND -DCUSTOM_MUTATOR_BOLT11 -fsanitize=address,fuzzer -Wall -Wextra -std=c++20 -I include -I .  main.cpp modules/ldk/module.a modules/lnd/module.a driver.o include/bitcoinfuzz/basemodule.o -o bitcoinfuzz 
main.cpp:65:10: error: no member named 'memcpy' in namespace 'std'
   65 |     std::memcpy(fuzz_data, dummy.data(), output_size);
      |     ~~~~~^
main.cpp:57:56: warning: unused parameter 'size' [-Wunused-parameter]
   57 | static size_t initial_input(uint8_t *fuzz_data, size_t size, size_t max_size)
      |                                                        ^
main.cpp:182:10: error: no member named 'memcpy' in namespace 'std'
  182 |     std::memcpy(fuzz_data, output.data(), output.length());
      |     ~~~~~^
main.cpp:197:10: error: no member named 'memcpy' in namespace 'std'
  197 |     std::memcpy(out, in1, insert_begin);
      |     ~~~~~^
main.cpp:198:10: error: no member named 'memcpy' in namespace 'std'
  198 |     std::memcpy(out + insert_begin, in2 + in2_begin, insert_size);
      |     ~~~~~^
main.cpp:199:10: error: no member named 'memcpy' in namespace 'std'
  199 |     std::memcpy(out + insert_begin + insert_size, in1 + insert_begin,
      |     ~~~~~^
main.cpp:216:10: error: no member named 'memcpy' in namespace 'std'
  216 |     std::memcpy(out, in1, in1_size);
      |     ~~~~~^
main.cpp:217:10: error: no member named 'memcpy' in namespace 'std'
  217 |     std::memcpy(out + overwrite_begin, in2 + in2_begin, overwrite_size);
      |     ~~~~~^
main.cpp:305:10: error: no member named 'memcpy' in namespace 'std'
  305 |     std::memcpy(out, encoded.data(), encoded.size());
      |     ~~~~~^
1 warning and 8 errors generated.
make: *** [Makefile:16: bitcoinfuzz] Error 1
```

Adding this include fixes the errors.
`#include <cstring> `

2. After including cstring and trying to compile i got this error:

```
make
clang++ -DLDK -DLND -DCUSTOM_MUTATOR_BOLT11 -fsanitize=address,fuzzer -Wall -Wextra -std=c++20 -I include -I .  main.cpp modules/ldk/module.a modules/lnd/module.a driver.o include/bitcoinfuzz/basemodule.o -o bitcoinfuzz 
main.cpp:58:56: warning: unused parameter 'size' [-Wunused-parameter]
   58 | static size_t initial_input(uint8_t *fuzz_data, size_t size, size_t max_size)
      |                                                        ^
1 warning generated.
/usr/bin/ld: /tmp/main-4a10ba.o: in function `LLVMFuzzerCustomMutator':
main.cpp:(.text.LLVMFuzzerCustomMutator[LLVMFuzzerCustomMutator]+0x4f1): undefined reference to `bech32::Decode(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bech32::CharLimit)'
/usr/bin/ld: main.cpp:(.text.LLVMFuzzerCustomMutator[LLVMFuzzerCustomMutator]+0x1b12): undefined reference to `bech32::Encode(bech32::Encoding, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<unsigned char, std::allocator<unsigned char> > const&)'
/usr/bin/ld: /tmp/main-4a10ba.o: in function `LLVMFuzzerCustomCrossOver':
main.cpp:(.text.LLVMFuzzerCustomCrossOver[LLVMFuzzerCustomCrossOver]+0x597): undefined reference to `bech32::Decode(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bech32::CharLimit)'
/usr/bin/ld: main.cpp:(.text.LLVMFuzzerCustomCrossOver[LLVMFuzzerCustomCrossOver]+0x6b6): undefined reference to `bech32::Decode(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bech32::CharLimit)'
/usr/bin/ld: main.cpp:(.text.LLVMFuzzerCustomCrossOver[LLVMFuzzerCustomCrossOver]+0x1158): undefined reference to `bech32::Encode(bech32::Encoding, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<unsigned char, std::allocator<unsigned char> > const&)'
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [Makefile:16: bitcoinfuzz] Error 1
```

3. When trying to build the bitcoin core I am receiving this error:
```
CMake Error in CMakeLists.txt:
  No cmake_minimum_required command is present.  A line of code such as

    cmake_minimum_required(VERSION 4.0)

  should be added at the top of the file.  The version specified may be lower
  if you wish to support older CMake versions for this project.  For more
  information run "cmake --help-policy CMP0000".


-- Configuring incomplete, errors occurred!
make: *** [Makefile:35: ../univalue/build/libunivalue.a] Error 1
```

Similar to the one that appeared on this ci: https://github.com/brunoerg/bitcoinfuzz/actions/runs/14203350756/job/39795216201#step:11:228

erickcestari commented:
Other thing that I noticed is that our ci is not recompiling the bitcoinfuzz for each test in CI. #128

brunoerg commented:
I will put on the docs that to use this custom mutator it's required to build the Bitcoin Core module for now. So fixing the Bitcoin Core module build should work, I'm checking it. 

It's weird that this build error wasn't occurring before, since we did not change anything on this? (?)

brunoerg commented:
rebased

brunoerg committed:
ci: remove nlightning since there is a known issue

morehouse commented on main.cpp:
```diff
@@ -38,8 +39,275 @@
+    const std::string dummy =
+        "lnbc16lta047pp5h6lta047h6lta047h6lta047h6lta047h6lta047h6lta047"
+	    "h6lqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqq"
+	    "qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqxnht6w";
```
Nit: this can be a compile-time constant:

```suggestion
    constexpr std::string_view dummy =
        "lnbc16lta047pp5h6lta047h6lta047h6lta047h6lta047h6lta047h6lta047"
	    "h6lqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqq"
	    "qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqxnht6w";
```

brunoerg replied:
Done

morehouse commented on main.cpp:
```diff
@@ -38,8 +39,275 @@
+
 std::shared_ptr<bitcoinfuzz::Driver> driver = nullptr;
 
+#ifdef CUSTOM_MUTATOR_BOLT11
```
Can this whole `ifdef` be moved to `modules/bitcoin/bech32.h` or somewhere else, so that `main.cpp` doesn't get polluted?

brunoerg replied:
I'm thinking about a better place to move it to. I agree that `main.cpp` isn't the most elegant place for it.

morehouse commented on main.cpp:
```diff
@@ -38,8 +39,275 @@
+//      `max_size`, otherwise return the encoded dummy invoice.
+extern "C" size_t LLVMFuzzerMutate(uint8_t *Data, size_t Size, size_t MaxSize);
+
+size_t LLVMFuzzerCustomMutator(uint8_t *fuzz_data, size_t size, size_t max_size,
+                               unsigned int seed)
```
I'm not super familiar with all the license rules, but this mutator is mostly copied from CLN.  Probably bitcoinfuzz needs an open source license in order to use this code...

It's probably also good to credit CLN as the original source of the mutator.

brunoerg replied:
I just added a license here and credited CLN for this mutator.

morehouse commented on main.cpp:
```diff
@@ -38,8 +39,275 @@
+//      dummy on failure).
+//   4. Write the encoded result to `fuzz_data` if its size does not exceed
+//      `max_size`, otherwise return the encoded dummy invoice.
+extern "C" size_t LLVMFuzzerMutate(uint8_t *Data, size_t Size, size_t MaxSize);
```
Nit: the `LLVMFuzzerMutate` definition should be moved further up so it doesn't separate the `LLVMFuzzerCustomMutator` definition from its function comment.

brunoerg replied:
Done

morehouse commented on main.cpp:
```diff
@@ -38,8 +39,275 @@
+    auto data = decoded.data;
+    auto hrp = decoded.hrp;
+
+    size_t data_maxlen = input.size() > 8 ? std::min(input.size() - 8, static_cast<size_t>(65)) : 0;
+    size_t hrp_maxlen = input.size() > 6 ? std::min(input.size() - 6, static_cast<size_t>(83)) : 0;
```
Nit:  IMO brace initialization is cleaner than `static_cast`:

```suggestion
    size_t data_maxlen = input.size() > 8 ? std::min(input.size() - 8, size_t{65}) : 0;
    size_t hrp_maxlen = input.size() > 6 ? std::min(input.size() - 6, size_t{83}) : 0;
```

morehouse commented on main.cpp:
```diff
@@ -38,8 +39,275 @@
+                hrp_buffer.resize(hrp_maxlen);
+
+                // Mutate the buffer
+                size_t new_len = LLVMFuzzerMutate(hrp_buffer.data(),
+                                                 hrp.size(),
+                                                 hrp_maxlen - 1);
```
Can we use `hrp_maxlen` without subtracting 1?  IIRC CLN's mutator left an extra byte to add the null terminator, which is needed for CLN's parsing.  But I don't think bitcoinfuzz needs that null terminator.

brunoerg replied:
Yes, done.

morehouse commented on main.cpp:
```diff
@@ -38,8 +39,275 @@
+                                                 hrp_maxlen - 1);
+
+                // Ensure minimum length and validate - at least 1 character
+                new_len = std::max(new_len, static_cast<size_t>(1));
```
```suggestion
                new_len = std::max(new_len, size_t{1});
```

brunoerg replied:
Done

morehouse commented on main.cpp:
```diff
@@ -38,8 +39,275 @@
+                                                 hrp.size(),
+                                                 hrp_maxlen - 1);
+
+                // Ensure minimum length and validate - at least 1 character
```
If libFuzzer truncates the HRP or data part to length 0, why do we ignore that and enforce len 1 instead?

brunoerg replied:
I've changed this part and removed it.

morehouse commented on main.cpp:
```diff
@@ -38,8 +39,275 @@
+        case 1: { // Mutate data
+            if (data_maxlen > 0) {
+                // Make sure we have a buffer that's large enough for mutation
+                data.resize(std::max(data.size(), data_maxlen));
+
+                size_t datalen = data.size();
+                size_t new_len = LLVMFuzzerMutate(data.data(),
+                                                 std::min(datalen, data_maxlen),
+                                                 data_maxlen);
```
I don't think this is doing the right thing...

1. `data` get's resized to `data_maxlen`.
2. `datalen` gets set to `data_maxlen`.
3. `LLVMFuzzerMutate` is called with `data_maxlen`.

Shouldn't `LLVMFuzzerMutate` be called with the original `data` size, not the resized size?

brunoerg replied:
Yes, you're right. I just fixed it.

morehouse commented on main.cpp:
```diff
@@ -38,8 +39,275 @@
+                                                 std::min(datalen, data_maxlen),
+                                                 data_maxlen);
+
+                new_len = std::max(new_len, static_cast<size_t>(1));
```
```suggestion
                new_len = std::max(new_len, size_t{1});
```

brunoerg replied:
Done

morehouse commented on main.cpp:
```diff
@@ -38,8 +39,275 @@
+    auto data = decoded.data;
+    auto hrp = decoded.hrp;
+
+    size_t data_maxlen = input.size() > 8 ? std::min(input.size() - 8, static_cast<size_t>(65)) : 0;
+    size_t hrp_maxlen = input.size() > 6 ? std::min(input.size() - 6, static_cast<size_t>(83)) : 0;
```
Where do the 65 and 83 magic constants come from?

brunoerg replied:
It's from the BIP173 and other implementations but feel free to ignore, I just removed them.

morehouse commented on main.cpp:
```diff
@@ -38,8 +39,275 @@
+        return initial_input(fuzz_data, size, max_size);
+    } else if (!hrp.empty()) {
+        for (const char& c : hrp) {
+            if (c >= 'A' && c <= 'Z') {
+                return initial_input(fuzz_data, size, max_size);
+            }
```
How do we get in this case, given that we enforce lowercase above?

brunoerg replied:
We enforce only if we mutate the hrp part, right? If it takes the path to mutate the data, the HRP is left unchanged, and if it originally contained uppercase letters, those aren't sanitized.

morehouse commented on main.cpp:
```diff
@@ -38,8 +39,275 @@
+            hrp1_vec.data(), hrp1_vec.size(),
+            hrp2_vec.data(), hrp2_vec.size(),
+            out_hrp_vec.data(), out_hrp_vec.size(),
+            (unsigned)std::rand());
```
static_cast?

brunoerg replied:
Done

brunoerg commented:
Thanks, @morehouse. Just addressed your points.

brunoerg committed:
add bolt11mutator module

This module has the bech32 functions from Bitcoin Core
to be used by the BOLT11 custom mutator.

brunoerg commented:
I just added a module called `bolt11mutator`. It contains only the bech32 files from Bitcoin Core, very light and simple. This way we don't need to compile the entire Bitcoin Core module to be able to use the bech32 stuff for the mutator.

morehouse commented on .github/workflows/workflow.yml:
```diff
@@ -52,7 +52,11 @@ jobs:
+
+      - name: Install libsodium
+        run: |
+          sudo apt install libsodium-dev
```
I don't see libsodium required anywhere in this PR.

brunoerg replied:
I've dropped it. It was a workaround to test a thing.

morehouse commented on .github/workflows/workflow.yml:
```diff
@@ -146,19 +155,18 @@ jobs:
 
       - name: Test - deserialize_invoice
         run: |
-          export CXXFLAGS="-DLDK -DLND -DNLIGHTNING"
```
Why is `-DNLIGHTNING` removed?

brunoerg replied:
Because there is a known issue on nlightning and with the custom mutator, 100 runs were enough to catch it and cause a crash.

morehouse commented on main.cpp:
```diff
@@ -38,8 +39,271 @@
+
 std::shared_ptr<bitcoinfuzz::Driver> driver = nullptr;
 
+#ifdef CUSTOM_MUTATOR_BOLT11
```
Can we move this to the new `bolt11mutator` module?

brunoerg replied:
Sure, it's my goal. I will do it in a follow-up since I've addressed the other points and I'm running out of time for now.

morehouse commented on modules/bitcoin/bech32.h:
```diff
@@ -37,6 +37,7 @@ enum class Encoding {
  *  and we would never encode an address with such a massive value */
 enum CharLimit : size_t {
     BECH32 = 90,            //!< BIP173/350 imposed character limit for Bech32(m) encoded addresses. This guarantees finding up to 4 errors.
+    CUSTOM_MUTATOR = 100'000,
```
Can we remove this now that there's a separate module for the custom mutator?

brunoerg replied:
Sure. I'm going to remove it.

morehouse commented on main.cpp:
```diff
@@ -38,8 +39,271 @@
+    std::srand(seed);
+    switch (std::rand() % 2) {
+        case 0: { // Mutate hrp
+            if (hrp_maxlen > 0) {
```
I don't think it's possible for `hrp_maxlen` to be 0. We short circuit if `size < 9`, so the min value for `hrp_maxlen` is 3.

brunoerg replied:
I just removed this check.

morehouse commented on main.cpp:
```diff
@@ -38,8 +39,271 @@
+                                                 data_maxlen);
+
+                new_len = std::max(new_len, size_t{1});
+                data.resize(new_len);
```
This in-place resize and mutation of `data` should also be possible for `hrp` -- `std::string` allows resizing too.

morehouse commented on main.cpp:
```diff
@@ -38,8 +39,271 @@
+            break;
+        }
+        case 1: { // Mutate data
+            if (data_maxlen > 0) {
```
AFAICT `data_maxlen` is always at least 1 here, so we can remove this check.

brunoerg replied:
Done.

morehouse commented on main.cpp:
```diff
@@ -38,8 +39,271 @@
+                                                 original_data_size,
+                                                 data_maxlen);
+
+                new_len = std::max(new_len, size_t{1});
```
Can we remove this as well, like we did for the HRP?

brunoerg replied:
Done. I removed for both.

morehouse commented on main.cpp:
```diff
@@ -38,8 +39,271 @@
+
+                for (auto& val : data) {
+                    val &= 0x1F;
+                }
```
A comment would be helpful here.

brunoerg replied:
Done.

morehouse commented on main.cpp:
```diff
@@ -38,8 +39,271 @@
+        }
+    }
+
+    if (hrp.empty() || hrp.size() > 83) {
```
We should not be enforcing the BIP 173 limits.  BOLT 11 explicitly allows larger invoices:  https://github.com/lightning/bolts/blob/master/11-payment-encoding.md#requirements

brunoerg replied:
Yes, you're right. I just removed it.

morehouse commented on main.cpp:
```diff
@@ -38,8 +39,271 @@
+    }
+
+    if (hrp.empty() || hrp.size() > 83) {
+        return initial_input(fuzz_data, size, max_size);
```
Can we just allow an empty HRP or data part to see how implementations handle it?

brunoerg replied:
Sounds good, I removed it.

morehouse commented on main.cpp:
```diff
@@ -38,8 +39,271 @@
+
+    std::string output = bech32::Encode(bech32::Encoding::BECH32, hrp, data);
+
+    if (output.empty() || output.length() > max_size) {
```
I don't think it's possible for `output` to be empty here.

brunoerg replied:
Sure, I just removed this check.

morehouse commented on main.cpp:
```diff
@@ -38,8 +39,271 @@
+        out_data = data1;
+
+        // Convert strings to vectors for cross_over function
+        std::vector<uint8_t> hrp1_vec(hrp1.begin(), hrp1.end());
+        std::vector<uint8_t> hrp2_vec(hrp2.begin(), hrp2.end());
```
Do the HRPs actually need to be converted?  Maybe we can just use `data()` and `size()` on the strings themselves.

We might also be able to do this for `out_hrp`.

brunoerg replied:
I removed the conversion and it's now using `std::span` for the cross_over.

morehouse commented on main.cpp:
```diff
@@ -38,8 +39,271 @@
+        // Cross-over the data part
+        out_hrp = hrp1;
+
+        size_t max_out_data_size = max_out_size - hrp1.size() - 8;
+        std::vector<uint8_t> out_data_vec(max_out_data_size);
+
+        size_t out_data_size = cross_over(
+            data1.data(), data1.size(),
+            data2.data(), data2.size(),
+            out_data_vec.data(), max_out_data_size,
+            (unsigned)std::rand());
+
+        out_data.assign(out_data_vec.begin(), out_data_vec.begin() + out_data_size);
```
We can use `out_data` directly.  Also should use `static_cast`:

```suggestion
        out_data.resize(max_out_size - hrp1.size() - 8);

        size_t out_data_size = cross_over(
            data1.data(), data1.size(),
            data2.data(), data2.size(),
            out_data.data(), out_data.size(),
            static_cast<unsigned>(std::rand()));

        out_data.resize(out_data_size);
```

brunoerg replied:
I've changed cross_over and simplified this part a little bit.

morehouse commented on main.cpp:
```diff
@@ -38,8 +39,271 @@
+    // Encode the output
+    std::string encoded = bech32::Encode(bech32::Encoding::BECH32, out_hrp, out_data);
+
+    if (encoded.empty() || encoded.size() > max_out_size) {
```
I don't think `encoded` can be empty here.

brunoerg replied:
I just removed this check.

morehouse commented on main.cpp:
```diff
@@ -38,8 +39,271 @@
+    return in1_size;
+}
+
+size_t cross_over(const uint8_t *in1, size_t in1_size, const uint8_t *in2,
```
This function can be `static`.  Or better yet, all these helpers can go in an unnamed namespace.

brunoerg replied:
Just made it `static` but I intend to move it to the `bolt11mutator` module.

morehouse commented on main.cpp:
```diff
@@ -38,8 +39,271 @@
+    return in1_size;
+}
+
+size_t cross_over(const uint8_t *in1, size_t in1_size, const uint8_t *in2,
+                size_t in2_size, uint8_t *out, size_t max_out_size, unsigned seed)
```
Nit:  these parameters could be C++-ified using `std::span` for inputs and `std::vector` for the output.

brunoerg replied:
Done.

morehouse commented on main.cpp:
```diff
@@ -38,8 +40,290 @@
+
+    if (data.empty() || data.size() > 65) {
+        return initial_input(fuzz_data, max_size);
+    }
```
We shouldn't enforce the 65 limit, and I don't see a strong need to require `data` to be non-empty either.

```suggestion
```

brunoerg replied:
Yes, it makes sense, I've removed it.

morehouse commented on main.cpp:
```diff
@@ -38,8 +40,290 @@
+    // Choose overwrite position
+    if (!out.empty()) {
+        size_t pos = std::rand() % out.size();
+        size_t overwrite_length = std::min(in2.size(), out.size() - pos);
```
This will *always* overwrite the max amount from `in2`.  We still want some randomness here so that a substring of `in2` could be overwritten instead.

brunoerg replied:
Just changed it to select a random substring.

morehouse commented on main.cpp:
```diff
@@ -38,8 +40,290 @@
+        size_t overwrite_length = std::min(in2.size(), out.size() - pos);
+
+        // Overwrite portion of out with in2
+        std::copy_n(in2.begin(), overwrite_length, out.begin() + pos);
```
This will always start overwriting with the beginning of `in2`, which is also undesired.  We want some random substring of `in2` to be used.

brunoerg replied:
Done.

morehouse commented on main.cpp:
```diff
@@ -38,8 +40,290 @@
+    if (out.size() < max_out_size) {
+        size_t remaining = max_out_size - out.size();
+        size_t in2_bytes = std::min(in2.size(), remaining);
+        out.insert(out.end(), in2.begin(), in2.begin() + in2_bytes);
```
This *always* inserts the maximum substring of `in2`.  We want to insert some random substring instead.

brunoerg replied:
Done.

morehouse commented on main.cpp:
```diff
@@ -38,8 +40,290 @@
+        // Cross-over the HRP
+        out_data = data1;
+
+        std::span<const uint8_t> hrp1_span(reinterpret_cast<const uint8_t*>(hrp1.data()), hrp1.size());
+        std::span<const uint8_t> hrp2_span(reinterpret_cast<const uint8_t*>(hrp2.data()), hrp2.size());
```
This is uglier than I thought.  I was hoping we could just pass the string directly to `cross_over` and it would get implicitly converted to a span.

But I guess the issue is that `char` doesn't convert implicitly to `uint8_t`?

morehouse reviewed (changes requested):
The modified `insert_part` and `overwrite_part` functions have behavior that is too restrictive, so they won't produce as many unique combinations as the original mutator from CLN.

morehouse commented on main.cpp:
```diff
@@ -38,8 +40,289 @@
+    out.insert(out.end(), in1.begin(), in1.begin() + in1_bytes);
+
+    // Choose overwrite position
+    if (!out.empty() && !in2.empty()) {
```
At this point `in2` can't be empty, or we would have returned above.

And we could reduce nesting by inverting the condition:

```c++
if (out.empty()) return out;
```

brunoerg replied:
Done

morehouse commented on main.cpp:
```diff
@@ -38,8 +40,289 @@
+    if (!out.empty() && !in2.empty()) {
+        size_t pos = std::rand() % out.size();
+        size_t max_possible = std::min(in2.size(), out.size() - pos);
+        size_t overwrite_length = max_possible > 0 ? (std::rand() % max_possible) + 1 : 0;
```
AFAICT `max_possible` cannot be 0 here.

We already checked that `in2.size() > 0`, and `out.size() - pos` is always at least 1.

brunoerg replied:
Done

morehouse commented on main.cpp:
```diff
@@ -38,8 +40,289 @@
+        size_t max_possible = std::min(in2.size(), out.size() - pos);
+        size_t overwrite_length = max_possible > 0 ? (std::rand() % max_possible) + 1 : 0;
+
+        if (overwrite_length > 0) {
```
Likewise, since we added 1 above, `overwrite_length` is always at least 1.

brunoerg replied:
Done.

morehouse reviewed (approved):
Code LGTM, haven't tested.
